### PR TITLE
[C++] Project schema via field Ids and Schema intersection

### PR DESCRIPTION
**Pull request by eddyxu**
- Add `Schema.Project()` over a list of field IDs(int32)
- Add `Schema.Intersection(const Schema&)`

### CODE REVIEW

File: cpp/src/lance/format/schema.cc [hunk 1]
```diff
@@ -690,12 +784,13 @@ std::vector<lance::format::pb::Field> Schema::ToProto() const {
   return pb_fields;
 }
 
-/// Make a full copy of the schema.
+/// Make a full deep copy of the schema.
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:787]: make a note on what "deep" means here?

File: cpp/src/lance/format/schema.h [hunk 2]
```diff
@@ -35,17 +35,20 @@ class Field;
...
 class Schema final {
  public:
+  /// Field ID type.
+  using FieldIdType = int32_t;
+
   Schema() = default;
 
   /// Construct Lance Schema from Arrow Schema.
-  Schema(const std::shared_ptr<::arrow::Schema>& schema);
+  explicit Schema(const std::shared_ptr<::arrow::Schema>& schema);
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.h:44]: curious - why do we need to prevent implicit conversion here?

**Review by changhiskhan — APPROVED** [review]
minor questions

File: cpp/src/lance/format/schema.h (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.h:44]: This is a C++ style guide , and reported by clang-tidy

https://google.github.io/styleguide/cppguide.html#Implicit_Conversions

File: cpp/src/lance/format/schema_test.cc [hunk 3]
```diff
@@ -119,9 +122,48 @@ TEST_CASE("Get projection via arrow schema") {
   CHECK(expect_schema.Equals(projection, false));
 }
 
+TEST_CASE("Project using field ids") {
+  auto schema = lance::format::Schema(arrow_schema);
+  for (const auto& name : arrow_schema->field_names()) {
+    auto fields = schema.Project({name}).ValueOrDie()->GetFieldIds();
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/format/schema_test.cc:128]: Fixed

File: cpp/src/lance/format/schema.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc]: Done

**Pull request merged**